drivers/usb/serial/xr_usb_serial_hal.c

```diff
@@ -0,0 +1,855 @@
+	newctrl = xr_usb_serial->ctrlout;
+
+	set = (set & TIOCM_DTR ? XR_USB_SERIAL_CTRL_DTR : 0) | (set & TIOCM_RTS ? XR_USB_SERIAL_CTRL_RTS : 0);
+
+	clear = (clear & TIOCM_DTR ? XR_USB_SERIAL_CTRL_DTR : 0) | (clear & TIOCM_RTS ? XR_USB_SERIAL_CTRL_RTS : 0);
+
+	newctrl = (newctrl & ~clear) | set;
+
+	if (xr_usb_serial->ctrlout == newctrl)
+		return 0;
+
+	xr_usb_serial->ctrlout = newctrl;
+
+	if (newctrl & XR_USB_SERIAL_CTRL_DTR)
+		xr_usb_serial_set_reg(xr_usb_serial, xr_usb_serial->reg_map.uart_gpio_clr_addr, 0x08);
+	else
+		xr_usb_serial_set_reg(xr_usb_serial, xr_usb_serial->reg_map.uart_gpio_set_addr, 0x08);
+
+	if (newctrl & XR_USB_SERIAL_CTRL_RTS)
+		xr_usb_serial_set_reg(xr_usb_serial, xr_usb_serial->reg_map.uart_gpio_clr_addr, 0x20);
+	else
+		xr_usb_serial_set_reg(xr_usb_serial, xr_usb_serial->reg_map.uart_gpio_set_addr, 0x20);
+
+	return 0;
+}
+
+static struct reg_addr_map xr21b140x_reg_map;
+static struct reg_addr_map xr21b1411_reg_map;
+static struct reg_addr_map xr21v141x_reg_map;
+static struct reg_addr_map xr21b142x_reg_map;
+
+static void init_xr21b140x_reg_map(void)
+{
+	xr21b140x_reg_map.uart_enable_addr = 0x00;
+	xr21b140x_reg_map.uart_format_addr = 0x05;
+	xr21b140x_reg_map.uart_flow_addr = 0x06;
+	xr21b140x_reg_map.uart_loopback_addr = 0x16;
+	xr21b140x_reg_map.uart_xon_char_addr = 0x07;
+	xr21b140x_reg_map.uart_xoff_char_addr = 0x08;
+	xr21b140x_reg_map.uart_gpio_mode_addr = 0x0c;
+	xr21b140x_reg_map.uart_gpio_dir_addr = 0x0d;
+	xr21b140x_reg_map.uart_gpio_set_addr = 0x0e;
+	xr21b140x_reg_map.uart_gpio_clr_addr = 0x0f;
+	xr21b140x_reg_map.uart_gpio_status_addr = 0x10;
+	xr21b140x_reg_map.tx_break_addr = 0x0a;
+	xr21b140x_reg_map.uart_custom_driver = 0x41;
+}
+
+static void init_xr21b1411_reg_map(void)
+{
+	xr21b1411_reg_map.uart_enable_addr = 0xc00;
+	xr21b1411_reg_map.uart_flow_addr = 0xc06;
+	xr21b1411_reg_map.uart_loopback_addr = 0xc16;
+	xr21b1411_reg_map.uart_xon_char_addr = 0xc07;
+	xr21b1411_reg_map.uart_xoff_char_addr = 0xc08;
+	xr21b1411_reg_map.uart_gpio_mode_addr = 0xc0c;
+	xr21b1411_reg_map.uart_gpio_dir_addr = 0xc0d;
+	xr21b1411_reg_map.uart_gpio_set_addr = 0xc0e;
+	xr21b1411_reg_map.uart_gpio_clr_addr = 0xc0f;
+	xr21b1411_reg_map.uart_gpio_status_addr = 0xc10;
+	xr21b1411_reg_map.tx_break_addr = 0xc0a;
+	xr21b1411_reg_map.uart_custom_driver = 0x20d;
+}
+
+static void init_xr21v141x_reg_map(void)
+{
+	xr21v141x_reg_map.uart_enable_addr = 0x03;
+	xr21v141x_reg_map.uart_format_addr = 0x0b;
+	xr21v141x_reg_map.uart_flow_addr = 0x0c;
+	xr21v141x_reg_map.uart_loopback_addr = 0x12;
+	xr21v141x_reg_map.uart_xon_char_addr = 0x10;
+	xr21v141x_reg_map.uart_xoff_char_addr = 0x11;
+	xr21v141x_reg_map.uart_gpio_mode_addr = 0x1a;
+	xr21v141x_reg_map.uart_gpio_dir_addr = 0x1b;
+	xr21v141x_reg_map.uart_gpio_set_addr = 0x1d;
+	xr21v141x_reg_map.uart_gpio_clr_addr = 0x1e;
+	xr21v141x_reg_map.uart_gpio_status_addr = 0x1f;
+	xr21v141x_reg_map.tx_break_addr = 0x14;
+}
+
+static void init_xr21b142x_reg_map(void)
```

Comment:
if all these register maps are fixed, probably you don't need a function for initialize each one of them... they can be declared as const and initialized in the header (xr_usb_serial_common.h)....